pkg/csvcopy/transaction.go

```diff
@@ -0,0 +1,226 @@
+	err := conn.QueryRowxContext(ctx, statement, tr.loc.ImportID, tr.loc.StartRow, tr.loc.RowCount, tr.loc.ByteOffset, tr.loc.ByteLen).Scan(&retry)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return ErrTransactionAlreadyCompleted
+		}
+		return err
+	}
+	return nil
+}
+
+func (tr Transaction) setFailed(ctx context.Context, conn sqlx.ExecerContext, reason string) error {
+	sql := `
+	INSERT INTO timescaledb_parallel_copy.transactions (
+		import_id, start_row, row_count, byte_offset, byte_len,
+		created_at, state, failure_reason
+	)
+	VALUES ($1, $2, $3, $4, $5, NOW(), 'failed', $6)
+	ON CONFLICT (import_id, start_row)
+	-- If there was a previous failure, update it
+	DO UPDATE SET
+		state = 'failed',
+		failure_reason = $6,
+		created_at = NOW()
+	WHERE transactions.state = 'failed'
+	`
+	_, err := conn.ExecContext(ctx, sql, tr.loc.ImportID, tr.loc.StartRow, tr.loc.RowCount, tr.loc.ByteOffset, tr.loc.ByteLen, reason)
+	return err
+}
+
+// Get returns the row stats for the current transaction
+func (tr Transaction) Get(ctx context.Context, conn sqlx.QueryerContext) (*TransactionRow, error) {
+	return getTransactionRow(ctx, conn, tr.loc.ImportID, tr.loc.StartRow)
+}
+
+func getTransactionRow(ctx context.Context, conn sqlx.QueryerContext, fileID string, startRow int64) (*TransactionRow, error) {
+	row := &TransactionRow{}
+
+	err := conn.QueryRowxContext(ctx, `
+		SELECT import_id, start_row, row_count, byte_offset, byte_len, created_at, state, failure_reason
+		FROM timescaledb_parallel_copy.transactions
+		WHERE import_id = $1 AND start_row = $2
+		LIMIT 1
+	`, fileID, startRow).Scan(
+		&row.ImportID, &row.StartRow, &row.RowCount, &row.ByteOffset, &row.ByteLen, &row.CreatedAt, &row.State, &row.FailureReason,
+	)
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, nil
+	}
+	if err != nil {
+		return row, err
+	}
+
+	return row, nil
+}
+
+// Next returns the Next transaction in the sequence.
+// If it returns nil, it means there is no Next transaction
+func (tr Transaction) Next(ctx context.Context, conn sqlx.QueryerContext) (*Transaction, *TransactionRow, error) {
+	row := TransactionRow{}
+
+	err := conn.QueryRowxContext(ctx, `
+		SELECT import_id, start_row, row_count, byte_offset, byte_len, created_at, state, failure_reason
+		FROM timescaledb_parallel_copy.transactions
+		WHERE import_id = $1 AND start_row > $2
+		ORDER BY start_row ASC
+		LIMIT 1
+	`, tr.loc.ImportID, tr.loc.StartRow).Scan(
+		&row.ImportID, &row.StartRow, &row.RowCount, &row.ByteOffset, &row.ByteLen, &row.CreatedAt, &row.State, &row.FailureReason,
+	)
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return nil, nil, nil
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	next := newTransactionAt(Location{
+		ImportID:   row.ImportID,
+		StartRow:   row.StartRow,
+		RowCount:   row.RowCount,
+		ByteOffset: row.ByteOffset,
+		ByteLen:    row.ByteLen,
+	})
+
+	return next, &row, nil
+}
+
+func ensureTransactionTable(ctx context.Context, connString string) error {
```

Comment:
I think we should provide a command to clean up the objects we create in the DB.